server.js

```diff
@@ -2087,21 +2093,30 @@ if (config.startServer) {
           }
         };
 
-        // If we're running in EC2 auto scaling, this will wait for our instance
-        // to enter the `Terminating:Wait` state.
-        //
-        // If we're not running in EC2, this will be a no-op.
-        lifecycleHooks
-          .handleAndCompleteInstanceTermination(prepareForTermination)
-          .catch((err) => {
-            logger.error('Error handling instance termination', err);
+        // SIGTERM can be used to gracefully shut down the process. This signal
+        // may come from another process, but we also send it to ourselves if
+        // we want to gracefully shut down. This is used below in the ASG
+        // lifecycle handler, and also within the "terminate" webhook.
+        process.once('SIGTERM', async () => {
+          await prepareForTermination();
+
+          try {
+            await lifecycleHooks.completeInstanceTermination();
+          } catch (err) {
+            logger.error('Error completing instance termination', err);
             Sentry.captureException(err);
-          })
-          .finally(terminate);
+          }
+
+          await terminate();
```

Comment:
I guess we really don't want `terminate()` to throw an exception? It all looks safe with the possible exception of `Sentry.captureException()` in the catch -- maybe this should also be inside a `try` so that it can't possibly throw?

Reply:
That's probably safest, yeah. Though something is probably extremely broken if that ever throws!